From: =?UTF-8?q?Mika=20Pfl=C3=BCger?= <debian@mikapflueger.de>
Date: Mon, 27 Feb 2012 01:43:26 +0100
Subject: Debian nagios file locations and additional policy

This addresses distribution specific file locations, allows the nagios mail
plugin to run sudo for the purpose of directly searching postfix spools (more
policy needed for other MTAs), allows nrpe to create tmp files and do all
networking stuff, and to talk to database servers.
---
 policy/modules/admin/sudo.if            |   26 +++++++++++++++
 policy/modules/kernel/corenetwork.te.in |    3 +-
 policy/modules/services/nagios.fc       |    9 ++++-
 policy/modules/services/nagios.te       |   52 ++++++++++++++++++++++++++++++-
 4 files changed, 86 insertions(+), 4 deletions(-)

diff --git a/policy/modules/admin/sudo.if b/policy/modules/admin/sudo.if
index 975af1a..3c76c74 100644
--- a/policy/modules/admin/sudo.if
+++ b/policy/modules/admin/sudo.if
@@ -177,3 +177,29 @@ interface(`sudo_sigchld',`
 
 	allow $1 sudodomain:process sigchld;
 ')
+
+#######################################
+## <summary>
+##	Execute sudo_exec_t without a domain transition
+## </summary>
+## <desc>
+##	<p>
+##	This interface allows a domain to execute sudo_exec_t without a
+##	domain transition.  It is for daemons that already have setuid
+##	access but are running as uid != 0.
+##	</p>
+## </desc>
+## <param name="domain">
+##	<summary>
+##	The domain that can execute sudo.
+##	</summary>
+## </param>
+#
+template(`can_exec_sudo',`
+
+	gen_require(`
+		type sudo_exec_t;
+	')
+
+	can_exec($1, sudo_exec_t)
+')
diff --git a/policy/modules/kernel/corenetwork.te.in b/policy/modules/kernel/corenetwork.te.in
index daa10aa..d6c51f5 100644
--- a/policy/modules/kernel/corenetwork.te.in
+++ b/policy/modules/kernel/corenetwork.te.in
@@ -119,7 +119,7 @@ network_port(http, tcp,80,s0, tcp,443,s0, tcp,488,s0, tcp,8008,s0, tcp,8009,s0,
 network_port(http_cache, udp,3130,s0, tcp,8080,s0, tcp,8118,s0, tcp,10001-10010,s0) # 8118 is for privoxy
 network_port(i18n_input, tcp,9010,s0)
 network_port(imaze, tcp,5323,s0, udp,5323,s0)
-network_port(inetd_child, tcp,1,s0, udp,1,s0, tcp,7,s0, udp,7,s0, tcp,9,s0, udp,9,s0, tcp,13,s0, udp,13,s0, tcp,19,s0, udp,19,s0, tcp,37,s0, udp,37,s0, tcp,512,s0, tcp,543,s0, tcp,544,s0, tcp,891,s0, udp,891,s0, tcp,892,s0, udp,892,s0, tcp,2105,s0, tcp,5666,s0)
+network_port(inetd_child, tcp,1,s0, udp,1,s0, tcp,7,s0, udp,7,s0, tcp,9,s0, udp,9,s0, tcp,13,s0, udp,13,s0, tcp,19,s0, udp,19,s0, tcp,37,s0, udp,37,s0, tcp,512,s0, tcp,543,s0, tcp,544,s0, tcp,891,s0, udp,891,s0, tcp,892,s0, udp,892,s0, tcp,2105,s0)
 network_port(innd, tcp,119,s0)
 network_port(ipmi, udp,623,s0, udp,664,s0)
 network_port(ipp, tcp,631,s0, udp,631,s0, tcp,8610-8614,s0, udp,8610-8614,s0)
@@ -156,6 +156,7 @@ network_port(netsupport, tcp,5404,s0, udp,5404,s0, tcp,5405,s0, udp,5405,s0)
 network_port(nmbd, udp,137,s0, udp,138,s0)
 network_port(ntop, tcp,3000-3001,s0, udp,3000-3001,s0)
 network_port(ntp, udp,123,s0)
+network_port(nrpe, tcp,5666,s0)
 network_port(oracledb, tcp, 1521,s0,udp, 1521,s0, tcp,2483,s0,udp,2483,s0, tcp,2484,s0, udp,2484,s0)
 network_port(ocsp, tcp,9080,s0)
 network_port(openvpn, tcp,1194,s0, udp,1194,s0)
diff --git a/policy/modules/services/nagios.fc b/policy/modules/services/nagios.fc
index 1fc9905..4d2ea62 100644
--- a/policy/modules/services/nagios.fc
+++ b/policy/modules/services/nagios.fc
@@ -1,7 +1,9 @@
 /etc/nagios(/.*)?					gen_context(system_u:object_r:nagios_etc_t,s0)
 /etc/nagios/nrpe\.cfg				--	gen_context(system_u:object_r:nrpe_etc_t,s0)
+ifndef(`distro_debian', `
 /etc/rc\.d/init\.d/nagios			--	gen_context(system_u:object_r:nagios_initrc_exec_t,s0)
 /etc/rc\.d/init\.d/nrpe				--	gen_context(system_u:object_r:nagios_initrc_exec_t,s0)
+')
 
 /usr/s?bin/nagios				--	gen_context(system_u:object_r:nagios_exec_t,s0)
 /usr/s?bin/nrpe					--	gen_context(system_u:object_r:nrpe_exec_t,s0)
@@ -17,9 +19,12 @@
 /var/spool/nagios(/.*)?					gen_context(system_u:object_r:nagios_spool_t,s0)
 
 ifdef(`distro_debian',`
-/usr/sbin/nagios				--	gen_context(system_u:object_r:nagios_exec_t,s0)
-')
+/usr/sbin/nagios.*				--	gen_context(system_u:object_r:nagios_exec_t,s0)
+/usr/lib/cgi-bin/nagios.?/.+   --      gen_context(system_u:object_r:nagios_cgi_exec_t,s0)
+/usr/lib/nagios3/cgi/.+        --      gen_context(system_u:object_r:nagios_cgi_exec_t,s0)
+', `
 /usr/lib(64)?/cgi-bin/nagios(/.+)?			gen_context(system_u:object_r:httpd_nagios_script_exec_t,s0)
+')
 /usr/lib(64)?/nagios/cgi-bin(/.*)?			gen_context(system_u:object_r:httpd_nagios_script_exec_t,s0)
 
 # admin plugins
diff --git a/policy/modules/services/nagios.te b/policy/modules/services/nagios.te
index bf64a4c..e72e552 100644
--- a/policy/modules/services/nagios.te
+++ b/policy/modules/services/nagios.te
@@ -143,6 +143,7 @@ optional_policy(`
 #
 # Nagios CGI local policy
 #
+apache_script_exec_domain(nagios)
 optional_policy(`
 	apache_content_template(nagios)
 	typealias httpd_nagios_script_t alias nagios_cgi_t;
@@ -191,8 +192,15 @@ read_files_pattern(nrpe_t, nagios_etc_t, nagios_etc_t)
 files_search_etc(nrpe_t)
 
 manage_files_pattern(nrpe_t, nrpe_var_run_t, nrpe_var_run_t)
+manage_files_pattern(nrpe_t, nagios_var_run_t, nagios_var_run_t)
 files_pid_filetrans(nrpe_t, nrpe_var_run_t, file)
 
+type nrpe_tmp_t;
+files_tmp_file(nrpe_tmp_t)
+manage_dirs_pattern(nrpe_t, nrpe_tmp_t, nrpe_tmp_t)
+manage_files_pattern(nrpe_t, nrpe_tmp_t, nrpe_tmp_t)
+files_tmp_filetrans(nrpe_t, nrpe_tmp_t, { file dir })
+
 kernel_read_system_state(nrpe_t)
 kernel_read_kernel_sysctls(nrpe_t)
 
@@ -202,6 +210,16 @@ corecmd_exec_shell(nrpe_t)
 corenet_tcp_bind_generic_node(nrpe_t)
 corenet_tcp_bind_inetd_child_port(nrpe_t)
 corenet_sendrecv_unlabeled_packets(nrpe_t)
+corenet_all_recvfrom_unlabeled(nrpe_t)
+corenet_all_recvfrom_netlabel(nrpe_t)
+corenet_tcp_sendrecv_all_if(nrpe_t)
+corenet_tcp_sendrecv_all_nodes(nrpe_t)
+corenet_tcp_sendrecv_generic_port(nrpe_t)
+corenet_tcp_bind_all_nodes(nrpe_t)
+corenet_tcp_bind_nrpe_port(nrpe_t)
+sysnet_dns_name_resolve(nrpe_t)
+
+allow nrpe_t self:netlink_route_socket create_netlink_socket_perms;
 
 dev_read_sysfs(nrpe_t)
 dev_read_urand(nrpe_t)
@@ -223,6 +241,15 @@ miscfiles_read_localization(nrpe_t)
 
 userdom_dontaudit_use_unpriv_user_fds(nrpe_t)
 
+domain_read_all_domains_state(nrpe_t)
+fs_getattr_all_fs(nrpe_t)
+storage_getattr_fixed_disk_dev(nrpe_t)
+init_read_utmp(nrpe_t)
+
+term_dontaudit_getattr_all_user_ttys(nrpe_t)
+term_dontaudit_getattr_unallocated_ttys(nrpe_t)
+term_dontaudit_getattr_all_user_ptys(nrpe_t)
+
 optional_policy(`
 	inetd_tcp_service_domain(nrpe_t, nrpe_exec_t)
 ')
@@ -270,6 +297,7 @@ files_getattr_all_file_type_fs(nagios_admin_plugin_t)
 #
 
 allow nagios_mail_plugin_t self:capability { setuid setgid dac_override };
+dontaudit nagios_mail_plugin_t self:capability { sys_resource };
 
 allow nagios_mail_plugin_t self:netlink_route_socket r_netlink_socket_perms;
 allow nagios_mail_plugin_t self:tcp_socket create_stream_socket_perms;
@@ -289,17 +317,25 @@ logging_send_syslog_msg(nagios_mail_plugin_t)
 
 sysnet_read_config(nagios_mail_plugin_t)
 
+files_read_usr_files(nagios_mail_plugin_t)
+
 optional_policy(`
 	mta_send_mail(nagios_mail_plugin_t)
 ')
 
 optional_policy(`
+	can_exec_sudo(nagios_mail_plugin_t)
+')
+
+optional_policy(`
 	nscd_dontaudit_search_pid(nagios_mail_plugin_t)
 ')
 
 optional_policy(`
 	postfix_stream_connect_master(nagios_mail_plugin_t)
-	posftix_exec_postqueue(nagios_mail_plugin_t)
+	posftix_run_postqueue(nagios_mail_plugin_t)
+	postfix_list_spool(nagios_mail_plugin_t)
+	postfix_read_spool_files(nagios_mail_plugin_t)
 ')
 
 ######################################
@@ -313,6 +349,7 @@ allow nagios_checkdisk_plugin_t self:capability { sys_admin sys_rawio };
 files_read_etc_runtime_files(nagios_checkdisk_plugin_t)
 
 fs_getattr_all_fs(nagios_checkdisk_plugin_t)
+files_getattr_all_mountpoints(nagios_checkdisk_plugin_t)
 
 storage_raw_read_fixed_disk(nagios_checkdisk_plugin_t)
 
@@ -343,6 +380,8 @@ optional_policy(`
 ')
 
 optional_policy(`
+	mysql_read_config(nagios_services_plugin_t)
+	mysql_tcp_connect(nagios_services_plugin_t)
 	mysql_stream_connect(nagios_services_plugin_t)
 ')
 
@@ -389,3 +428,14 @@ optional_policy(`
 optional_policy(`
 	unconfined_domain(nagios_unconfined_plugin_t)
 ')
+
+optional_policy(`
+        mysql_tcp_connect(nrpe_t)
+        mysql_stream_connect(nrpe_t)
+       mysql_read_config(nrpe_t)
+')
+
+optional_policy(`
+        postgresql_tcp_connect(nrpe_t)
+        postgresql_stream_connect(nrpe_t)
+')
